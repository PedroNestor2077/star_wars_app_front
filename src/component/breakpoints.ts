export const breakpoints = {
    mobile: '(max-width: 550px)',
    desktop: '(min-width: 550px)',
};
