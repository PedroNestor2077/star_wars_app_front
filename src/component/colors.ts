const colors = {
    grey1: '#333333',
    grey2: '#666666',
    grey3: '#757575',
    grey4: '#C8C8C8',
    grey5: '#E6E6E6',
    grey6: '#969696',
    black: '#000000',
    blue1: '#002B53',
    blue2: '#002B56',
};

export default colors;
