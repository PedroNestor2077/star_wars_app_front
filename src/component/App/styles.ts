import styled from 'styled-components';
import { breakpoints } from '../breakpoints';

export const App = styled.div`
    min-height: 100vh;
    min-width: calc(100vw - 100px);
    max-width: 100vw;
    padding-top: 80px;
    padding-bottom: 80px;
    margin-left: 50px;
    margin-right: 50px;

    @media ${breakpoints.mobile} {
        margin-left: 25px;
        margin-right: 25px;
    }
`;

export const InlineContainer = styled.div`
    display: flex;
    align-items: center;
    justify-content: flex-start;
    width: 100%;

    @media ${breakpoints.mobile} {
        justify-content: space-between;
    }
`;
