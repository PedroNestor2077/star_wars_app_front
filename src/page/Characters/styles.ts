import styled from 'styled-components';
import colors from '../../component/colors';
import { breakpoints } from '../../component/breakpoints';

export const Nav = styled.nav`
    display: flex;
    align-items: center;
    width: calc(100vw - 50px);
    padding-top: 26px;
    padding-bottom: 26px;
    padding-left: 50px;
    margin-left: -50px;
    border-top: 1px solid ${colors.grey5};
    border-bottom: 1px solid ${colors.grey5};

    @media ${breakpoints.mobile} {
        padding-left: 25px;
        margin-left: -25px;
        width: calc(100vw - 25px);
        padding-top: 16px;
        padding-bottom: 16px;
        border-bottom: none;
    }
`;

export const NavContainer = styled.nav`
    width: 100%;
    display: flex;
    align-items: center;
    padding-right: 50px;

    @media ${breakpoints.mobile} {
        padding-right: 25px;
    }
`;

export const LoadMoreContainer = styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    margin-top: 110px;

    @media ${breakpoints.mobile} {
    }
`;
