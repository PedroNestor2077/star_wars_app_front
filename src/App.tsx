import { App as StyledApp } from './component/App';
import './index.css';
import { Characters } from './page/Characters/Characters';

function App() {
    return (
        <StyledApp>
            <Characters />
        </StyledApp>
    );
}

export default App;
