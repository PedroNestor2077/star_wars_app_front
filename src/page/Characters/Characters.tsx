import { useEffect, useState } from 'react';
import { InlineContainer } from '../../component/App';
import { Button } from '../../component/Button';
import { PeopleListContainer } from '../../component/PersonList';
import { Select } from '../../component/Select';
import { Spacing } from '../../component/Spacing';
import { Label, MainTitle, Paragraph, Title } from '../../component/Typograph';
import { PEOPLE_URL, PLANETS_URL } from '../../constants';
import { LoadMoreContainer, Nav, NavContainer } from './styles';
import PeopleApi from '../../service/people/people.service';
import { People } from '../../component/Person';
import { PlanetData } from '../../service/planets/types';
import PlanetsApi from '../../service/planets/planets.service';
import { SwapiGateway } from '../../gateway/swapi/swapi.gateway';

type Filter = string | null;

export const Characters = () => {
    const [filter, setFilter] = useState<Filter>(null);
    const [peopleUrl, setPeopleUrl] = useState<string | null>(PEOPLE_URL);
    const [people, setPeople] = useState<People[]>([]);
    const [planets, setPlanets] = useState<PlanetData[]>([]);
    const [isLoading, setIsLoading] = useState<boolean>(false);

    const _swapiGateway = new SwapiGateway();
    const peopleService = new PeopleApi(_swapiGateway);
    const planetsService = new PlanetsApi(_swapiGateway);

    const getPlanets = async (url: string) => {
        setIsLoading(true);
        const result = await planetsService.getPlanets(url);
        setPlanets(result);
    };

    const getPeoples = async (url: string) => {
        setIsLoading(true);
        const result = await peopleService.getPeople(url, planets);
        setPeopleUrl(result?.nextUrl || null);
        setPeople(result.data);
        setIsLoading(false);
    };

    const filterPeople = (people: People[], filter: Filter) => {
        if (!filter) return people;

        const filteredData = people.filter(
            ({ homeland }) => homeland === filter
        );
        return filteredData;
    };

    const handleLoadMore = async () => {
        setIsLoading(true);
        if (!peopleUrl || !planets.length) return;
        const result = await peopleService.getPeople(peopleUrl, planets);
        setPeopleUrl(result?.nextUrl || null);
        setPeople([...people, ...result.data]);
        setIsLoading(false);
    };

    const hasResults = !(!filterPeople(people, filter)?.length && !isLoading);

    useEffect(() => {
        if (!planets.length) return;
        getPeoples(peopleUrl || PEOPLE_URL);
    }, [planets]);

    useEffect(() => {
        getPlanets(PLANETS_URL);
    }, []);

    return (
        <>
            <MainTitle>Star Wars Characters</MainTitle>
            <Spacing top={25} mobileTop={10} bottom={50} mobileBottom={28}>
                <Paragraph maxWidth={934}>
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed
                    do eiusmod tempor incididunt ut labore et dolore magna
                    aliqua.
                </Paragraph>
            </Spacing>

            <Nav>
                <NavContainer>
                    <InlineContainer>
                        <Spacing right={10}>
                            <Label>Filter By:</Label>
                        </Spacing>
                        <Select
                            options={planets.map(({ name }) => ({
                                label: name,
                                value: name,
                            }))}
                            placeholder="All"
                            onSelect={setFilter}
                            maxWidth={190}
                            value={filter}
                        />
                    </InlineContainer>
                    <Button
                        maxWidth={158}
                        hideInMobile
                        onClick={() => setFilter(null)}
                        disabled={!filter}
                    >
                        CLEAR ALL
                    </Button>
                </NavContainer>
            </Nav>

            <Spacing top={50} bottom={43} mobileTop={28} mobileBottom={28}>
                <Title>All Characters</Title>
            </Spacing>
            <>
                {hasResults && (
                    <PeopleListContainer>
                        {filterPeople(people, filter)?.map((people: People) => (
                            <People {...people} key={people.name} />
                        ))}
                    </PeopleListContainer>
                )}
                {!hasResults && <Label>No Results</Label>}
            </>

            <LoadMoreContainer>
                <Spacing mobileRight={72} mobileLeft={72} fullWidth>
                    {!isLoading && (
                        <Button
                            maxWidth={486}
                            onClick={handleLoadMore}
                            disabled={!peopleUrl}
                        >
                            LOAD MORE
                        </Button>
                    )}
                    {isLoading && <Label>LOADING...</Label>}
                </Spacing>
            </LoadMoreContainer>
        </>
    );
};
