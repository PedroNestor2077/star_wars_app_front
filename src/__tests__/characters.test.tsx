import { Characters } from '../page/Characters/Characters';
import { App } from '../component/App';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'; // Para adicionar expectativas especÃ­ficas do Jest-Dom

describe('Characters Page', () => {
    test('Should match snapshot', () => {
        const { container } = render(
            <App>
                <Characters />
            </App>
        );
        expect(container).toMatchSnapshot();
    });

    test('Should match snapshot (mobile)', () => {
        window.innerWidth = 500;

        const { container } = render(
            <App>
                <Characters />
            </App>
        );
        expect(container).toMatchSnapshot();

        window.innerWidth = 1024;
    });
});
