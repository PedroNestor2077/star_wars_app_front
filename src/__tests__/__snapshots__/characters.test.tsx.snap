// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters Page Should match snapshot (mobile) 1`] = `
<div>
  <div
    class="sc-beySPh ihIVNw"
  >
    <div
      class="sc-fsYfdN sc-qZrbh CFNGT fSJQZJ"
    >
      Star Wars Characters
    </div>
    <div
      bottom="50"
      class="sc-iBdnpw dQJYDs"
      mobilebottom="28"
      mobiletop="10"
      top="25"
    >
      <div
        class="sc-fsYfdN sc-kFCroH cMJWGl hCaBYj"
        maxwidth="934"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </div>
    </div>
    <nav
      class="sc-csKJxZ jJLqlh"
    >
      <nav
        class="sc-eTNRI cnjtxk"
      >
        <div
          class="sc-guDLey djQIhZ"
        >
          <div
            class="sc-iBdnpw fxJevY"
            right="10"
          >
            <div
              class="sc-fsYfdN sc-irLvIq CFNGT cliotF"
            >
              Filter By:
            </div>
          </div>
          <div
            class="sc-jTQCzO dykloI"
            data-testid="select-container"
            maxwidth="190"
          >
            <select
              class="sc-eDLKkx eMDarx"
              data-testid="select"
              disabled=""
            >
              <option
                value=""
              >
                All
              </option>
            </select>
            <p
              class="sc-gLLuof bDTxWV"
            >
              ⌄
            </p>
          </div>
        </div>
        <button
          class="sc-dmyCSP kluULB"
          disabled=""
          maxwidth="158"
        >
          CLEAR ALL
        </button>
      </nav>
    </nav>
    <div
      bottom="43"
      class="sc-iBdnpw hhuMHj"
      mobilebottom="28"
      mobiletop="28"
      top="50"
    >
      <div
        class="sc-fsYfdN sc-jsEeTM CFNGT cTplmx"
      >
        All Characters
      </div>
    </div>
    <div
      class="sc-hLQSwg gkBTae"
    />
    <div
      class="sc-brSamD cNiIRU"
    >
      <div
        class="sc-iBdnpw fspkaG"
        mobileleft="72"
        mobileright="72"
      >
        <div
          class="sc-fsYfdN sc-irLvIq CFNGT cliotF"
        >
          LOADING...
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Characters Page Should match snapshot 1`] = `
<div>
  <div
    class="sc-beySPh ihIVNw"
  >
    <div
      class="sc-fsYfdN sc-qZrbh CFNGT fSJQZJ"
    >
      Star Wars Characters
    </div>
    <div
      bottom="50"
      class="sc-iBdnpw dQJYDs"
      mobilebottom="28"
      mobiletop="10"
      top="25"
    >
      <div
        class="sc-fsYfdN sc-kFCroH cMJWGl hCaBYj"
        maxwidth="934"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </div>
    </div>
    <nav
      class="sc-csKJxZ jJLqlh"
    >
      <nav
        class="sc-eTNRI cnjtxk"
      >
        <div
          class="sc-guDLey djQIhZ"
        >
          <div
            class="sc-iBdnpw fxJevY"
            right="10"
          >
            <div
              class="sc-fsYfdN sc-irLvIq CFNGT cliotF"
            >
              Filter By:
            </div>
          </div>
          <div
            class="sc-jTQCzO dykloI"
            data-testid="select-container"
            maxwidth="190"
          >
            <select
              class="sc-eDLKkx eMDarx"
              data-testid="select"
              disabled=""
            >
              <option
                value=""
              >
                All
              </option>
            </select>
            <p
              class="sc-gLLuof bDTxWV"
            >
              ⌄
            </p>
          </div>
        </div>
        <button
          class="sc-dmyCSP kluULB"
          disabled=""
          maxwidth="158"
        >
          CLEAR ALL
        </button>
      </nav>
    </nav>
    <div
      bottom="43"
      class="sc-iBdnpw hhuMHj"
      mobilebottom="28"
      mobiletop="28"
      top="50"
    >
      <div
        class="sc-fsYfdN sc-jsEeTM CFNGT cTplmx"
      >
        All Characters
      </div>
    </div>
    <div
      class="sc-hLQSwg gkBTae"
    />
    <div
      class="sc-brSamD cNiIRU"
    >
      <div
        class="sc-iBdnpw fspkaG"
        mobileleft="72"
        mobileright="72"
      >
        <div
          class="sc-fsYfdN sc-irLvIq CFNGT cliotF"
        >
          LOADING...
        </div>
      </div>
    </div>
  </div>
</div>
`;
